### PR TITLE
Feature/rule 3.2

### PR DESCRIPTION
**Pull request by petertrr**
## Which rule and warnings did you add?

Rule 3.2 about general file structure, warnings: FILE_CONTAINS_ONLY_COMMENTS, FILE_INCORRECT_BLOCKS_ORDER, FILE_NO_BLANK_LINE_BETWEEN_BLOCKS, FILE_COLLAPSED_IMPORTS.

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [x] Added tests on fixers
* [x] Updated rules-config.json
* [x] Updated Readme
* [ ] Updated Wiki

## Fixme

* Custom ordering of imports is out of scope of this PR, but some baseline for it has been added.
* ~~Warning FILE_SINGLE_TOP_LEVEL_CLASS has been added but not implemented~~ this rule already exists as FILE_NAME_MATCH_CLASS

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -86,3 +86,8 @@ To install git hooks using gradle run `gradle installGitHooks`.
 |HEADER_MISSING_IN_NON_SINGLE_CLASS_FILE|Check: warns if file with zero or >1 classes doesn't have header KDoc|
 |COMMENTED_OUT_CODE|Check: warns if valid kotlin code is detected in commented blocks (both single-line and block comments)|
 |FILE_IS_TOO_LONG|Check: the number of lines in the parameter file is too high. Configuration has maxSize - maximum number of lines, ignoreFolders - files containing this folder in their path will not be checked
+|FILE_CONTAINS_ONLY_COMMENTS|Check: warns if file contains only comments, imports and package directive|
+|FILE_INCORRECT_BLOCKS_ORDER|Check: warns if general order of code parts is wrong. Fix: rearranges them.|
+|FILE_NO_BLANK_LINE_BETWEEN_BLOCKS|Check: warns if there is not exactly one blank line between code parts. Fix: leaves single empty line|
+|FILE_COLLAPSED_IMPORTS|Check: warns if collapsed imports are used.|
```
Comment (akuleshov7) [inline README.md:92]: you mean wildcard import?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+ * 5. Ensures there are no collapsed imports
+ */
+class FileStructureRule : Rule("file-structure") {
+
+    private lateinit var configRules: List<RulesConfig>
+    private lateinit var emitWarn: ((offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit)
+    private var isFixMode: Boolean = false
+    private var fileName: String = ""
+
+    override fun visit(node: ASTNode, autoCorrect: Boolean, params: KtLint.Params, emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:47]: let's have common style for visit function parameters. You have split lines in other Rules:
```
visit(node,
       autocorrect, 
```

File: README.md (hunk 1 above)
Comment (petertrr) [reply README.md:92]: Right, this is how they are called...

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        val codeTokens = TokenSet.andNot(TokenSet.ANY, TokenSet.create(WHITE_SPACE, KDOC, BLOCK_COMMENT, EOL_COMMENT, PACKAGE_DIRECTIVE, IMPORT_LIST))
+        val hasCode = node.getChildren(codeTokens).isNotEmpty()
+        if (!hasCode) {
+            FILE_CONTAINS_ONLY_COMMENTS.warn(configRules, emitWarn, isFixMode, fileName, node.startOffset)
+        }
+        return hasCode
+    }
+
+    private fun checkCodeBlocksOrderAndEmptyLines(node: ASTNode) {
+        val copyrightComment = node.findChildBefore(PACKAGE_DIRECTIVE, BLOCK_COMMENT)
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:72]: I have found out that there is my mistake in findChildBefore - I have added empty KDOC to it. Could you please add description?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+        // importPath can be null if import name cannot be parsed, which should be a very rare case, therefore !! should be safe here
+        imports.filter { (it.psi as KtImportDirective).importPath!!.isAllUnder }.forEach {
+            FILE_COLLAPSED_IMPORTS.warn(configRules, emitWarn, isFixMode, it.text, it.startOffset)
+        }
+
+        val sortedImports = imports.sortedBy { it.text }
+        if (sortedImports != imports) {
+            FILE_UNORDERED_IMPORTS.warnAndFix(configRules, emitWarn, isFixMode, fileName, node.startOffset) {
+                // move all commented lines among import before imports block
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:116]: please move fix block to a function

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    private fun checkImportsOrder(node: ASTNode) {
+        val imports = node.getChildren(TokenSet.create(IMPORT_DIRECTIVE)).toList()
+
+        // importPath can be null if import name cannot be parsed, which should be a very rare case, therefore !! should be safe here
+        imports.filter { (it.psi as KtImportDirective).importPath!!.isAllUnder }.forEach {
+            FILE_COLLAPSED_IMPORTS.warn(configRules, emitWarn, isFixMode, it.text, it.startOffset)
+        }
+
+        val sortedImports = imports.sortedBy { it.text }
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:113]: I forgot - we decided simply to order them alphabetically?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt (hunk 4 above)
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:116]: do we really want to move all comments to the top? what IDEA does for it?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        addChild(child.clone() as ASTNode, beforeThis)
+        if (nBlankLinesAfter > 0) addChild(PsiWhiteSpaceImpl("\n".repeat(nBlankLinesAfter)), beforeThis)
+        if (nBlankLinesAfter > 0) child.nextSibling { it.elementType == WHITE_SPACE }?.let { removeChild(it) }
+        removeChild(child)
+    }
+
+    // fixme recommended order (see Recommendation 3.1) of imports can be checked by custom comparator
+    private val recommendedComparator = Comparator<ASTNode> { import1, import2 ->
+        require(import1.elementType == IMPORT_DIRECTIVE && import2.elementType == IMPORT_DIRECTIVE) { "This comparator is for sorting imports" }
+        val segments1 = (import1.psi as KtImportDirective).importPath!!.fqName.pathSegments()
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:156]: what is segment1 and segment2?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                    if (index != 0) {
+                        node.addChild(PsiWhiteSpaceImpl("\n"), null)
+                    }
+                    node.addChild(importNode, null)
+                }
+            }
+        }
+    }
+
+    private fun ASTNode.getSiblingBlocks(copyrightComment: ASTNode?,
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:135]: I guess we can increase coverage if you would write tests for such functions like this

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt (hunk 5 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:113]: Yes, as a default sorting rule. When we write final version of more complicated sorting, it should be added as separate option.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt (hunk 4 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:116]: Yes, this is what IDEA does.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt (hunk 2 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:47]: We should add something like this to our code style workflow, probably as new check in diktat.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt (hunk 7 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt:135]: Specifically this method, imo, doesn't really need testing, it acts as a map with dynamic values. The next method, `ASTNode.moveChildBefore`, should be tested, but I actually have already implemented the same method with tests in #6. We should merge that PR first and I'll reuse that method here.

**Pull request merged**